rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -41,170 +41,284 @@ using Entity = liveliness::Entity;
-        // First time this topic name is added to the graph.
-        std::shared_ptr<TopicData> topic_data_ptr = std::make_shared<TopicData>(
-          topic_info,
-          TopicStats{pub_count, sub_count}
-        );
-        graph_endpoints[topic_info.name_] = GraphNode::TopicDataMap{
-          {topic_info.type_, topic_data_ptr}
-        };
+      // We check if an entity with the exact same qos also exists.
+      GraphNode::TopicQoSMap::iterator topic_qos_map_it = topic_type_map_it->second.find(
+        qos_str);
+      if (topic_qos_map_it == topic_type_map_it->second.end()) {
+        // First time this qos is added.
+        // Update cache.
+        topic_type_map_it->second.insert(std::make_pair(qos_str, graph_topic_data));
       } else {
-        // If a TopicData entry for the same type exists in the topic map, update pub/sub counts
-        // or else create an new TopicData.
-        std::pair<GraphNode::TopicDataMap::iterator, bool> topic_data_insertion =
-          cache_topic_it->second.insert(std::make_pair(topic_info.type_, nullptr));
-        if (topic_data_insertion.second) {
-          // A TopicData for the topic_type does not exist.
-          topic_data_insertion.first->second = std::make_shared<TopicData>(
-            topic_info,
-            TopicStats{pub_count, sub_count});
+        // We have another instance of a pub/sub over the same topic,
+        // type and qos so we increment the counters.
+        TopicDataPtr & existing_graph_topic = topic_qos_map_it->second;
+        if (is_pub) {
+          existing_graph_topic->pubs_.insert(entity);
         } else {
-          // Update the existing counters.
-          topic_data_insertion.first->second->stats_.pub_count_ += pub_count;
-          topic_data_insertion.first->second->stats_.sub_count_ += sub_count;
+          existing_graph_topic->subs_.insert(entity);
         }
       }
-    };
+    }
+  }
+  // Take events if any.
+  if (report_events) {
+    take_local_entities_with_events(local_entities_with_events);
+  }
+}
 
-  // Helper lambda to convert an Entity into a GraphNode.
-  // Note: this will update bookkeeping variables in GraphCache.
-  auto make_graph_node =
-    [&add_topic_data](const Entity & entity, GraphCache & graph_cache) -> std::shared_ptr<GraphNode>
-    {
-      auto graph_node = std::make_shared<GraphNode>();
-      graph_node->id_ = entity.id();
-      graph_node->ns_ = entity.node_namespace();
-      graph_node->name_ = entity.node_name();
-      graph_node->enclave_ = entity.node_enclave();
-
-      if (!entity.topic_info().has_value()) {
-        // Token was for a node.
-        return graph_node;
+///=============================================================================
+void GraphCache::take_local_entities_with_events(
+  std::unordered_map<liveliness::Entity, std::unordered_set<rmw_zenoh_event_type_t>> &
+  local_entities_with_events)
+{
+  if (local_entities_with_events.empty()) {
+    return;
+  }
```

Comment:
This doesn't seem to be strictly necessary; if there are no events, then we just won't enter into the loop below.

Reply:
Done https://github.com/ros2/rmw_zenoh/pull/103/commits/47973be62df1ea92f5387f9a7c5d5d5caec1e227